### ISSUE TITLE
`ref_patterns` vs match ergonomics

### ISSUE THREAD

**Issue by ronnodas** [rust-lang/rust-clippy#17725]
### Description

The scope of the `ref_patterns` lint is a little ambiguous from the description
> The `ref` keyword can be confusing for people unfamiliar with it, and often it is more concise to use `&` instead.

I can see at least two interpretations:
* Disallow any use of the `ref` keyword.
* Disallow uses of the `ref` keyword that can be replaced by match ergonomics, i.e. by inserting or removing some number of `&`/`&mut` on either side of the assignment.

In the second case the following is a false positive, adapted from the example in #16145:
```rust
fn foo(bar: &[u8]) -> u8 {
    if let &[a, b, ref rest @ .., c, d, e] = bar {
        baz([a, b, c, d, e], rest)
    } else {
        0
    }
}

fn baz(_: [u8; 5], _: &[u8]) -> u8 {
    0
}
```

In the intention is to disallow all uses of `ref`, then this code could be written as 
```rust
if let [a, b, rest @ .., c, d, e] = bar {
    baz([*a, *b, *c, *d, *e], rest)
}
```
but I don't see a way to have the same bindings using only match ergonomics. If that is the intended use of this lint, perhaps the description should be slightly amended.

### Version

```text

```
